### PR TITLE
e2e test design document

### PR DESCRIPTION
**Pull request by umezawatakeshi**

### CODE REVIEW

File: e2e/README.md [hunk 1]
```diff
@@ -0,0 +1,61 @@
+# End-to-end test suites
+
+## Strategy
+
+We try to maximize coverage of unit tests even if it needs "real" MySQL instances.  The tests using real MySQL instances can be divided into the unit tests and the e2e tests as follows:
+- test by unit test
```
Comment (ymmt2005) [inline e2e/README.md:6]: I'd suggest "small test" rather than unit test to denote the tests precisely.

File: e2e/README.md [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+- replication
+- backup/restore
+- point-in-time recovery
+- upgrade
+- blackout recovery
+- failure injection
+
+### YAML manifests
+
+`MySQLCluster` custom resource accepts some flexible settings such as the templates of Pod, Volume, and Service.  So, we should examine the settings are correctly applied to the resources made by MOCO.
```
Comment (ymmt2005) [inline e2e/README.md:42]: Such reconciliation results should be tests in small tests.
Does this document cover small tests in addition to e2e tests?

File: e2e/README.md [hunk 3]
```diff
@@ -0,0 +1,67 @@
+# End-to-end test suites
+
+## Strategy
+
+We try to maximize coverage of unit tests even if it needs "real" MySQL instances.
```
Comment (ymmt2005) [inline e2e/README.md:5]:
```suggestion
We try to maximize coverage of small tests even if it needs "real" MySQL instances.
```

File: e2e/README.md [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+The tests using real MySQL instances can be divided into the small tests and the e2e tests as follows:
```
Comment (ymmt2005) [inline e2e/README.md:6]:
```suggestion
Specifically, small tests should cover the following areas.
```

File: e2e/README.md [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+- by small test
+  - initialization of MySQL instance
+  - status retrieval
+  - cloning
+  - metrics collection
+  - log file management (rotate)
+- by e2e test
```
Comment (ymmt2005) [inline e2e/README.md:14]:
```suggestion
End-to-end (e2e) tests should include but not limited to:
```

File: e2e/README.md [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+  - failover
+  - switchover
+  - replication
+  - backup/restore
+  - point-in-time recovery
+  - upgrade
+  - blackout recovery
+  - failure injection
+
+The small tests are written in each `_test.go` file.
```
Comment (ymmt2005) [inline e2e/README.md:24]:
```suggestion
The small tests exist in each package directory.
```

File: e2e/README.md [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+The remains are:
```
Comment (ymmt2005) [inline e2e/README.md:26]: Why do these remain?  I guess these should be covered in e2e.

File: e2e/README.md [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+- YAML manifests
+- `main` functions of each programs
+
+## Analysis
+
+### Feature functions
```
Comment (ymmt2005) [inline e2e/README.md:32]: I'd propose "Scenarios" instead.

**Review by ymmt2005 — CHANGES_REQUESTED** [review]
Please fix and/or answer my questions.

File: e2e/README.md (hunk 2 above)
Comment (zoetrope) [reply e2e/README.md]: No. We will write about e2e tests.

**Review by ymmt2005 — APPROVED** [review]
LGTM

**Pull request merged**